agent/tests/Feature/IngestDetailsRepositoryTest.php

```diff
@@ -1,316 +1,358 @@
+        expect($loop)->toHaveRun([]);
+    }
+
+    public function test_it_handles_5xx_errors(): void
+    {
+        $loop = new LoopFake;
+        $browser = new BrowserFake([
+            Response::internalServerError('Whoops!'),
+        ]);
+
+        [$output, $e] = run(
+            via: 'source',
+            ingestDetailsBrowser: $browser,
+            loop: $loop,
+        );
+
+        expect($e)->toBeNull($e?->getMessage() ?? '');
+        expect($browser)->toHaveSent([
+            Request::json('/api/agent-auth'),
+        ]);
+        expect($browser)->toHavePending([]);
+        expect($output)->toMatchLog(<<<'OUTPUT'
         {date} {info} Authentication failed {duration}: 500 \[Whoops!\]
         OUTPUT);
-    expect($loop)->toHaveRun([]);
-});
-
-it('handles malformed JSON responses', function (string $body) {
-    $loop = new LoopFake;
-    $browser = new BrowserFake([
-        new Response($body),
-    ]);
-
-    [$output, $e] = run(
-        via: 'source',
-        ingestDetailsBrowser: $browser,
-        loop: $loop,
-    );
-
-    expect($e)->toBeNull($e?->getMessage() ?? '');
-    expect($browser)->toHaveSent([
-        Request::json('/api/agent-auth'),
-    ]);
-    expect($browser)->toHavePending([]);
-    expect($output)->toMatchLog(<<<'OUTPUT'
+        expect($loop)->toHaveRun([]);
+    }
+
+    #[DataProvider('malformedJson')]
+    public function test_it_handles_malformed_json_responses(string $body): void
+    {
+        $loop = new LoopFake;
+        $browser = new BrowserFake([
+            new Response($body),
+        ]);
+
+        [$output, $e] = run(
+            via: 'source',
+            ingestDetailsBrowser: $browser,
+            loop: $loop,
+        );
+
+        expect($e)->toBeNull($e?->getMessage() ?? '');
+        expect($browser)->toHaveSent([
+            Request::json('/api/agent-auth'),
+        ]);
+        expect($browser)->toHavePending([]);
+        expect($output)->toMatchLog(<<<'OUTPUT'
         {date} {info} Authentication failed {duration}: Syntax error
         OUTPUT);
-    expect($loop)->toHaveRun([]);
-})->with(['', '[']);
-
-it('handles unexpected response payloads', function (array $payload) {
-    $loop = new LoopFake;
-    $browser = new BrowserFake([
-        new Response($payload),
-    ]);
-
-    [$output, $e] = run(
-        via: 'source',
-        ingestDetailsBrowser: $browser,
-        loop: $loop,
-    );
-
-    expect($e)->toBeNull($e?->getMessage() ?? '');
-    expect($browser)->toHaveSent([
-        Request::json('/api/agent-auth'),
-    ]);
-    expect($browser)->toHavePending([]);
-    $payload = preg_quote(json_encode($payload, flags: JSON_THROW_ON_ERROR), '#');
-    expect($output)->toMatchLog(<<<OUTPUT
+        expect($loop)->toHaveRun([]);
+    }
+
+    /**
+     * @return iterable<array{0: string}>
+     */
+    public static function malformedJson(): iterable
+    {
+        yield [''];
+        yield ['['];
+        yield ['{'];
+    }
```

Comment:
I didn't know that the PHPUnit data provider method could return a generator!

Reply:
Big time!